software/iqtree/main.nf

```diff
@@ -0,0 +1,44 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+options        = initOptions(params.options)
+
+process IQTREE {
+    tag "$variant_alignment"
+    label 'process_medium'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), publish_id:'') }
+
+    conda (params.enable_conda ? "bioconda::iqtree=2.1.2" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/iqtree:2.1.2--h56fc30b_0"
+    } else {
+        container "quay.io/biocontainers/iqtree:2.1.2--h56fc30b_0"
+    }
+
+        input:
+    path variant_alignment
+    val constant_sites
+
+    output:
+    path "*.treefile", emit: phylogeny
+    path "*.version.txt", emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def fconst_args = constant_sites ? '-fconst $constant_sites' : ''
+    def memory = task.memory.toString().replaceAll(' ', '')
+    """
+    iqtree \\
+        $fconst_args \\
+        $options.args \\
+        -s $variant_alignment \\
+        -nt AUTO \\
```

Comment:
Should be coming in via `options.args`?

Reply:
These are sensible defaults

Reply:
Yup, it's not so much about having sensible defaults but rather allowing developer's to override as many module options as possible so they can run the module however they want. This is why we insist on only mandatory parameters being added here. Can always just add that to `modules.config` no?

Reply:
@drpatelh the args that are passed in https://github.com/nf-core/modules/blob/4bdb21eeba2ceb715e47bf3ec27962169a89abf8/software/iqtree/main.nf#L34-L40 are 
 * `$fconst_args` based on how the process is called (empty string or a string with numbers of constant sites)
 * `$options.args` obvs
 *  `-s $alignment` an input path
 * `-nt AUTO` which calculate the optimal number of cpus to use
 * `-ntmax $task.cpus` which ensures no more than the `$task.cpus` cpus are used
 * `-mem $memory` which makes sure that no more than the memory specified for the task is used 

I am not sure why a developer would need to override these. Doing so could mean that more cpus or memory than asked for are used.

Reply:
Ah, ok. My bad. I thought `-nt AUTO` was doing some sort of automated file format checking and not what you described above so yup, makes sense to have here 👍🏽